diff --git a/qemu/exec.c b/qemu/exec.c
index c753af6..1a135d8 100644
--- a/qemu/exec.c
+++ b/qemu/exec.c
@@ -719,6 +719,9 @@ static inline bool watchpoint_address_matches(CPUWatchpoint *wp,
 int cpu_watchpoint_address_matches(CPUState *cpu, vaddr addr, vaddr len)
 {
 #if ENABLE_CPU_WATCHPOINTS
+#ifdef TARGET_ARM
+    struct uc_struct *uc = cpu->uc;
+#endif
     CPUWatchpoint *wp;
     int ret = 0;
 
@@ -1325,7 +1328,8 @@ void cpu_check_watchpoint(CPUState *cpu, vaddr addr, vaddr len,
             cpu->watchpoint_hit = wp;
         }
     }
-    raise_exception_ra(cpu->env_ptr, EXCP_DEBUG, ra);
+    cpu->exception_index = EXCP_DEBUG;
+    cpu_loop_exit_restore(cpu, ra);
 #endif
 }
 
diff --git a/qemu/target/arm/translate.c b/qemu/target/arm/translate.c
index 744d8ff..e4e3dd0 100644
--- a/qemu/target/arm/translate.c
+++ b/qemu/target/arm/translate.c
@@ -18,6 +18,7 @@
  * You should have received a copy of the GNU Lesser General Public
  * License along with this library; if not, see <http://www.gnu.org/licenses/>.
  */
+#define GEN_ADAPTER_DEFINE
 #include "qemu/osdep.h"
 
 #include "cpu.h"
